### PR TITLE
Rewrite the analyze-sample-code script to also analyze snippets

### PR DESCRIPTION
**Pull request by gspencergoog**
This rewrites the sample code analysis script to be a little less of a hack (but still not pretty), and to handle snippets as well.

It also changes the semantics of how sample code is handled: the namespace for the sample code is now limited to the file that it appears in, so some additional "Examples can assume:" blocks were added.  The upside of this is that there will be far fewer name collisions.

I fixed the output too: no longer will you get 4000 lines of numbered output with the error at the top and have to grep for the actual problem.  It gives the filename and line number of the original location of the code (in the comment in the tree), and prints out the source code on the line that caused the problem along with the error.

For snippets, it prints out the location of the start of the snippet and the source code line that causes the problem.  It can't print out the original line, because snippets get formatted when they are written, so the line might not be in the same place.

I would actually like to integrate this more fully into the flutter command line tool so that we don't need to process the analyzer output, and can just use it directly, and ideally it would be able to do it in "continuous" mode, and we could get highlighting in the IDE directly.  That's a ways off though, so this will have to do until then.

### CODE REVIEW

**Comment by gspencergoog** [comment]
Note: I'd actually like to get this in on Monday, so that snippets will be syntax checked next week during the Doc fixit, so any help reviewing here would be great.  @tvolkert @Hixie

File: dev/bots/analyze-sample-code.dart [hunk 1]
```diff
@@ -38,174 +43,394 @@
...
+    _exitCode = 0;
+    Map<String, List<AnalysisError>> errors = <String, List<AnalysisError>>{};
+    try {
+      final Map<String, Section> sections = <String, Section>{};
+      final Map<String, Snippet> snippets = <String, Snippet>{};
+      _extractSamples(sections, snippets);
+      errors = _analyze(_tempDir, sections, snippets);
+    } finally {
+      if (errors.isNotEmpty) {
+        stderr.writeln('Kept temporary directory ${_tempDir.path} because there were errors.\n');
```
Comment (dnfield) [inline dev/bots/analyze-sample-code.dart]: Does it help to keep the directory in this case?  Once we've written the errors and line numbers out, can't we just go back to the original file?

Comment (dnfield) [reply dev/bots/analyze-sample-code.dart]: In particular Windows limits how many of these you can create before your process isn't allowed to create anymore

**Comment by dnfield** [comment]
Aside from the question about not cleaning up the temp dir on errors, this LGTM.

The analyzer output parsing is a little scary.  It'd be nice to get this as something that used the analyzer package directly instead  of parsing the output...

File: dev/bots/analyze-sample-code.dart (hunk 1 above)
Comment (gspencergoog) [reply dev/bots/analyze-sample-code.dart]: Good point, I'll remove that part and just always nuke the temp dir.

The Windows thing wouldn't matter: It only creates one temp dir.

**Pull request merged**